### DISCUSSION TITLE
Cost of cloning a stream

### DISCUSSION THREAD

**Discussion by davidbrochart** [agronholm/anyio#578] [Q&A] [answered]
Say you register a callback for sending to a stream, that will potentially be called a lot of times. The stream would need to be cloned, right?
```py
def synchronous_callback(send_stream: MemoryObjectSendStream) -> None:
    cloned = send_stream.clone()
    with cloned:
        cloned.send_nowait('hello')
```
I was wondering if cloning is expensive in terms of CPU, or negligible?
Or maybe I should not use a context manager and thus not close the stream in the first place?

**Comment by agronholm** [comment, marked as answer]
The cost is negligible. Memory object streams are supposed to be closed eagerly when they're no longer needed.

**Reply by agronholm** [reply]
Oh, one more thing: You're not handling the case when the buffer is full.

**Reply by agronholm** [reply]
Now that I look more closely, why are you cloning the stream in the first place?

**Reply by davidbrochart** [reply]
Yes that was one question, I guess I shouldn't use the context manager then, and just send.

**Reply by agronholm** [reply]
Yup. But what will you do if the queue is full?

**Reply by davidbrochart** [reply]
I can catch the `WouldBlock` exception and react accordingly. Do you mean the context manager helps in that regard?
